hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java

```diff
@@ -87,4 +106,125 @@ public <R> R run(Action<R, IMetaStoreClient, TException> action, boolean retry)
       throws TException, InterruptedException {
     return clientPool().run(action, retry);
   }
+
+  @VisibleForTesting
+  static Key toKey(List<Supplier<Object>> suppliers) {
+    return Key.of(suppliers.stream().map(Supplier::get).collect(Collectors.toList()));
+  }
+
+  @VisibleForTesting
+  static List<Supplier<Object>> extractKeySuppliers(String cacheKeys, Configuration conf) {
+    URIElement uri = URIElement.of(conf.get(HiveConf.ConfVars.METASTOREURIS.varname, ""));
+    if (cacheKeys == null || cacheKeys.isEmpty()) {
+      return Collections.singletonList(() -> uri);
+    }
+
+    // generate key elements in a certain order, so that the Key instances are comparable
+    Set<KeyElementType> types = Sets.newTreeSet(Comparator.comparingInt(Enum::ordinal));
+    Map<String, String> confElements = Maps.newTreeMap();
+    for (String element : cacheKeys.split(",", -1)) {
+      String trimmed = element.trim();
+      if (trimmed.toLowerCase(Locale.ROOT).startsWith(CONF_ELEMENT_PREFIX)) {
+        String key = trimmed.substring(CONF_ELEMENT_PREFIX.length());
+        ValidationException.check(
+            !confElements.containsKey(key), "Conf key element %s already specified", key);
+        confElements.put(key, conf.get(key));
+      } else {
+        KeyElementType type = KeyElementType.valueOf(trimmed.toUpperCase());
+        switch (type) {
+          case URI:
+          case UGI:
+          case USER_NAME:
+            ValidationException.check(
+                types.add(type), "%s key element already specified", type.name());
+            break;
+          default:
+            throw new ValidationException("Unknown key element %s", trimmed);
+        }
+      }
+    }
+    ImmutableList.Builder<Supplier<Object>> suppliers = ImmutableList.builder();
+    for (KeyElementType type : types) {
+      switch (type) {
+        case URI:
+          suppliers.add(() -> uri);
+          break;
+        case UGI:
+          suppliers.add(
+              () -> {
+                try {
+                  return UserGroupInformation.getCurrentUser();
+                } catch (IOException e) {
+                  throw new UncheckedIOException(e);
+                }
+              });
+          break;
+        case USER_NAME:
+          suppliers.add(
+              () -> {
+                try {
+                  String userName = UserGroupInformation.getCurrentUser().getUserName();
+                  return UserNameElement.of(userName);
+                } catch (IOException e) {
+                  throw new UncheckedIOException(e);
+                }
+              });
+          break;
+        default:
+          throw new RuntimeException("Unexpected key element " + type.name());
+      }
+    }
+    for (String key : confElements.keySet()) {
+      ConfElement element = ConfElement.of(key, confElements.get(key));
+      suppliers.add(() -> element);
+    }
+    return suppliers.build();
+  }
+
+  @Value.Immutable
```

Comment:
I'm strugling to see the value of adding these immutable value, although I may be missing something.  Why not just use String?

Reply:
It's meant to give different types for different keys. E.g. both URI and USER_NAME are essentially just a wrapper of a string, but they are not comparable and can never be considered equal to each other.

Reply:
The strings and lists are used in cache keys and need to be compared. I think we can either assume "how could a user name have the same value as an HMS URI", or we can wrap them in different classes to make sure they won't be equivalent. Personally I prefer the latter. But since we generally agree URI should be mandatory, I guess we can remove these two wrappers for now.